File: backend/candidate-profile-service/src/main/java/com/jobportal/candidateprofile/controllers/CandidateCertificateController.java
package com.jobportal.candidateprofile.controllers;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import com.jobportal.candidateprofile.dto.CandidateCertificateDto;
import com.jobportal.candidateprofile.service.CandidateCertificateService;


@RestController
@RequestMapping("/api/certificates")
@CrossOrigin("*")
public class CandidateCertificateController {

	@Autowired
	private CandidateCertificateService service;

	@PostMapping
	public CandidateCertificateDto addCertificate(@RequestBody CandidateCertificateDto dto) {
		return service.addCertificate(dto);
	}

	@PutMapping("/{certiId}")
	public CandidateCertificateDto updateCertificate(@PathVariable Integer certiId,
			@RequestBody CandidateCertificateDto dto) {
		return service.updateCertificate(certiId, dto);
	}

	@DeleteMapping("/{certiId}")
	public String deleteCertificate(@PathVariable Integer certiId) {
		service.deleteCertificate(certiId);
		return "Certificate deleted successfully";
	}

	@GetMapping("/{certiId}")
	public CandidateCertificateDto getCertificateById(@PathVariable Integer certiId) {
		return service.getCertificateById(certiId);
	}

	@GetMapping("/candidate/{cid}")
	public List<CandidateCertificateDto> getCertificatesByCandidate(@PathVariable Integer cid) {
		return service.getCertificatesByCandidate(cid);
	}

	@GetMapping
	public List<CandidateCertificateDto> getAllCertificates() {
		return service.getAllCertificates();
	}
}

File: backend-dotnet/Services/JobPortal.CandidateService/Controllers/CandidateCertificateController.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using JobPortal.CandidateService.DTOs;
using JobPortal.CandidateService.Services;

namespace JobPortal.CandidateService.Controllers;

[ApiController]
[Route("api/certificates")]
public class CandidateCertificateController : ControllerBase
{
    private readonly CandidateCertificateService _service;

    public CandidateCertificateController(CandidateCertificateService service)
    {
        _service = service;
    }

    [HttpPost]
    public async Task<IActionResult> AddCertificate([FromBody] CandidateCertificateDto dto)
    {
        var result = await _service.AddCertificateAsync(dto);
        return Ok(result);
    }

    [HttpPut("{certiId}")]
    public async Task<IActionResult> UpdateCertificate(int certiId, [FromBody] CandidateCertificateDto dto)
    {
        try
        {
            var result = await _service.UpdateCertificateAsync(certiId, dto);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Certificate not found");
        }
    }

    [HttpDelete("{certiId}")]
    public async Task<IActionResult> DeleteCertificate(int certiId)
    {
        try
        {
            await _service.DeleteCertificateAsync(certiId);
            return Ok("Certificate deleted successfully");
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Certificate not found");
        }
    }

    [HttpGet("{certiId}")]
    public async Task<IActionResult> GetCertificateById(int certiId)
    {
        try
        {
            var result = await _service.GetCertificateByIdAsync(certiId);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Certificate not found");
        }
    }

    [HttpGet("candidate/{cid}")]
    public async Task<IActionResult> GetCertificatesByCandidate(int cid)
    {
        var list = await _service.GetCertificatesByCandidateAsync(cid);
        return Ok(list);
    }

    [HttpGet]
    public async Task<IActionResult> GetAllCertificates()
    {
        var list = await _service.GetAllCertificatesAsync();
        return Ok(list);
    }
}
